rac2: send-queue tracking in push mode

sumeerbhola opened the pull request:
The new raftEventForReplica encapsulates all that is needed to handle the RaftEvent, including (re)creating a replicaSendStream. The complexity is encapsulated in the construction of this event, which also pays attention to regressions, jumps, and other anomalies, all of which cause the existing replicaSendStream (if any) to be recreated. The recreaton is a significant simplification from the prototype, which was doing unnecessarily complicated work in fixing up the state.

The send-queue tracking is simple, [indexToSend, nextRaftIndex). Since there is a send-queue, we separately track the eval.tokensDeducted counts. Tracker.UntrackGE is no longer needed, which is another simplification.

Epic: CRDB-37515

Informs #123509

Release note: None

sumeerbhola committed:
rac2: send-queue tracking in push mode

The new raftEventForReplica encapsulates all that is needed to handle the
RaftEvent, including (re)creating a replicaSendStream. The complexity is
encapsulated in the construction of this event, which also pays attention
to regressions, jumps, and other anomalies, all of which cause the
existing replicaSendStream (if any) to be recreated. The recreaton is a
significant simplification from the prototype, which was doing
unnecessarily complicated work in fixing up the state.

The send-queue tracking is simple, [indexToSend, nextRaftIndex). Since
there is a send-queue, we separately track the eval.tokensDeducted counts.
Tracker.UntrackGE is no longer needed, which is another simplification.

Epic: CRDB-37515

Informs #123509

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/130908)
<!-- Reviewable:end -->

sumeerbhola reviewed (commented):
I've fixed the existing `TestRangeController`, but there are other test failures, probably due to a combination of not providing the right input (FollowerStateInfo.Next and RaftEvent.MsgApps need to be properly populated) and due to code bugs. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/130908#-:-O714KAp7CbRrA9U6npM:b-nh3c65)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @kvoli and @pav-kv)



<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/130908)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @kvoli and @pav-kv)

___
*[`pkg/kv/kvserver/kvflowcontrol/rac2/range_controller.go` line 494 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/130908#-O714wXW5p1ZodJnlFAn:-O714wXW5p1ZodJnlFAo:b-oof4u1) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc3b0ea3fcab5349e7cf6f95003e6a24d8e3bd3a/pkg/kv/kvserver/kvflowcontrol/rac2/range_controller.go#L494)):*
> ```Go
> // RaftEvent.MsgApps on followerStataInfo.Next. msgApps is the map entry in
> // RaftEvent.MsgApps for this replica. For other parameters, see the struct
> // declarations.
> ```

This function will need a dedicated test -- all the complexity in the PR is here.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/130908)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @kvoli and @pav-kv)

___
*[`pkg/kv/kvserver/kvflowcontrol/rac2/testdata/simulation/range_controller_stream_connection` line 12 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/130908#-O74K1TkBRPEXrR5YNpN:-O74K1TkBRPEXrR5YNpO:bvgnob5) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c2c3ce76d06e42f092506f4286ea32f85fc43a8/pkg/kv/kvserver/kvflowcontrol/rac2/testdata/simulation/range_controller_stream_connection#L12)):*
> ```
> t=0s             handle=h op=connect    stream=t1/s1   log-position=4/0
> t=0s             handle=h op=connect    stream=t1/s2   log-position=4/0
> t=5s             handle=h op=connect    stream=t1/s3   log-position=4/50
> ```

To make the simulation test run, I've ripped out the log-position input. A stream that is "connecting" (in the old jargon) is a creation of a `replicaSendStream` and it should join at the current log position (quorumPosition + 1), so that future appends can construct the proper MsgApp for it. Specifying a particular log-position at t=5s when appends have been happening to the RangeController from 0-5s already is not easily workable. Also, these simulation tests are not really about having `replicaSendStreams` with different send-queues. For that, we'll want to extend the TestRangeController.


<!-- Sent from Reviewable.io -->

kvoli reviewed (commented):
> Reviewed 10 of 12 files at r1, 8 of 8 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/130908)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @pav-kv and @sumeerbhola)

___
*[`pkg/kv/kvserver/kvflowcontrol/rac2/range_controller.go` line 457 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/130908#-O74KnRZBB3V71SoRMmA:-O74KnRZBB3V71SoRMmB:b-htypa1) ([raw file](https://github.com/cockroachdb/cockroach/blob/e33f93f20859a5a7de8e3aec136a295830ac4d22/pkg/kv/kvserver/kvflowcontrol/rac2/range_controller.go#L457)):*
> ```Go
> 	// entry will be added.
> 	//
> 	// State and Match are the latest state. Next represents the state preceding
> ```

Is `State` here referring to `followerStateInfo`?
___
*[`pkg/kv/kvserver/kvflowcontrol/rac2/range_controller.go` line 524 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/130908#-O74ThD11i02DVSOocVo:-O74ThD11i02DVSOocVp:b-faeo0p) ([raw file](https://github.com/cockroachdb/cockroach/blob/e33f93f20859a5a7de8e3aec136a295830ac4d22/pkg/kv/kvserver/kvflowcontrol/rac2/range_controller.go#L524)):*
> ```Go
> 				if !initialized {
> 					msgAppFirstIndex = entry.Index
> 					msgAppLastIndex = msgAppFirstIndex + 1
> ```

nit: `msgAppLastIndex` is actually the last entry index + 1, the comment above `[msgAppFirstIndex, msgAppLastIndex)` is helpful but I found myself double checking this down below, consider renaming to something upper bound related.
___
*[`pkg/kv/kvserver/kvflowcontrol/rac2/range_controller.go` line 555 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/130908#-O74SoFB-Y4dFZ4DL1wJ:-O74SoFB-Y4dFZ4DL1wK:blrpqi5) ([raw file](https://github.com/cockroachdb/cockroach/blob/e33f93f20859a5a7de8e3aec136a295830ac4d22/pkg/kv/kvserver/kvflowcontrol/rac2/range_controller.go#L555)):*
> ```Go
> 		}
> 	} else {
> 		createSendStream = true
> ```

 Why is a send stream created when `!existingSendStreamState.existsAndInStateReplicate`? I thought we'd leave the send stream closed if not in `StateReplicate`?

EDIT: nvm, I see we're only calling this when in `StateReplicate` in `HandleRaftEvent...`.
___
*[`pkg/kv/kvserver/kvflowcontrol/rac2/testdata/simulation/range_controller_stream_connection` line 12 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/130908#-O74K1TkBRPEXrR5YNpN:-O74PvxIEVyVNN-rtKyE:brxp9sq) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c2c3ce76d06e42f092506f4286ea32f85fc43a8/pkg/kv/kvserver/kvflowcontrol/rac2/testdata/simulation/range_controller_stream_connection#L12)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

To make the simulation test run, I've ripped out the log-position input. A stream that is "connecting" (in the old jargon) is a creation of a `replicaSendStream` and it should join at the current log position (quorumPosition + 1), so that future appends can construct the proper MsgApp for it. Specifying a particular log-position at t=5s when appends have been happening to the RangeController from 0-5s already is not easily workable. Also, these simulation tests are not really about having `replicaSendStreams` with different send-queues. For that, we'll want to extend the TestRangeController.
</blockquote></details>

Sounds good.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/130908)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @kvoli and @pav-kv)

___
*[`pkg/kv/kvserver/kvflowcontrol/rac2/range_controller.go` line 457 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/130908#-O74KnRZBB3V71SoRMmA:-O74zaGc-Mj7dXCzF923:b-o5zg9i) ([raw file](https://github.com/cockroachdb/cockroach/blob/e33f93f20859a5a7de8e3aec136a295830ac4d22/pkg/kv/kvserver/kvflowcontrol/rac2/range_controller.go#L457)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

Is `State` here referring to `followerStateInfo`?
</blockquote></details>

Yep. Updated
___
*[`pkg/kv/kvserver/kvflowcontrol/rac2/range_controller.go` line 524 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/130908#-O74ThD11i02DVSOocVo:-O74zgXSD22QLTzuDmYM:bep1g8i) ([raw file](https://github.com/cockroachdb/cockroach/blob/e33f93f20859a5a7de8e3aec136a295830ac4d22/pkg/kv/kvserver/kvflowcontrol/rac2/range_controller.go#L524)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

nit: `msgAppLastIndex` is actually the last entry index + 1, the comment above `[msgAppFirstIndex, msgAppLastIndex)` is helpful but I found myself double checking this down below, consider renaming to something upper bound related.
</blockquote></details>

I'll leave this change to you when you take over this PR.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/130908)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @kvoli and @pav-kv)

___
*[`pkg/kv/kvserver/kvflowcontrol/rac2/range_controller.go` line 913 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/130908#-O7561hG74xEbVplZC4W:-O7561hG74xEbVplZC4X:b4sjyg8) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d65500cd04d9db25b23280964527d11b8a74d98/pkg/kv/kvserver/kvflowcontrol/rac2/range_controller.go#L913)):*
> ```Go
> 		// tracker. Which is clearly not viable.
> 		//
> 		// We have another issue when maintaining the send-queue in push-mode. We
> ```

I realized this when reviewing your PR that sets `RaftMessageRequest.UsingRAC2Protocol`. This is the best I could come up with.


<!-- Sent from Reviewable.io -->

kvoli commented:
Rebased on master and resolved conflict.

kvoli committed:
rac2: unit test construct raft event for replica

Add a unit test `TestConstructRaftEventForReplica`, which asserts on the
`raftEventForReplica` returned from `constructRaftEventForReplica`.

Note that one of the testcases, `.../msgapp entries beyond new entries`,
is commented out with a TODO, questioning whether the test input is
valid: `MsgApp`s with entries that are not included in
`raftEventAppendState`.

Part of: #130432
Release note: None

kvoli commented:
Pushed a unit test for `constructRaftEventForReplica`:

> **rac2: unit test construct raft event for replica**
>
>Add a unit test `TestConstructRaftEventForReplica`, which asserts on the
`raftEventForReplica` returned from `constructRaftEventForReplica`.
>
>Note that one of the testcases, `.../msgapp entries beyond new entries`,
is commented out with a TODO, questioning whether the test input is
valid: `MsgApp`s with entries that are not included in
`raftEventAppendState`.
>
>Part of: https://github.com/cockroachdb/cockroach/issues/130432
>Release note: None

kvoli committed:
rac2: rename msp app last index -> msg app ub index

Rename `msgAppLastIndex` to `msgAppUBIndex` within
`constructRaftEventForReplica`. The index represents a non-inclusive
upper bound for entries being sent:

```
[msgAppFirstIndex, msgAppUBIndex)
```
The rename makes it clearer that the index is not inclusive.

Part of: #130432
Release note: None

kvoli commented:
Pushed a rename discussed in [a prior comment](https://github.com/cockroachdb/cockroach/pull/130908#pullrequestreview-2313241525):


> rac2: rename msp app last index -> msg app ub index 
> 
> Rename `msgAppLastIndex` to `msgAppUBIndex` within
> `constructRaftEventForReplica`. The index represents a non-inclusive
> upper bound for entries being sent:
> 
> ```
> [msgAppFirstIndex, msgAppUBIndex)
> ```
> The rename makes it clearer that the index is not inclusive.
> 
> Part of: cockroachdb#130432
> Release note: None

kvoli committed:
rac2: add dd test for msg app raft events

Add a new `TestRangeController` variant `handle_raft_event_partial_msg_app`.

The test issues MsgApps in addition to the usual entries e.g.

```
raft_event
range_id=1
  entries
    term=1 index=1 pri=NormalPri size=1MiB
    term=1 index=2 pri=NormalPri size=1MiB
    term=1 index=3 pri=NormalPri size=1MiB
  sending
    replica_id=1 [1,4)
    replica_id=2 [1,3)
    replica_id=3 [1,2)
```
Part of: #130432
Release note: None

kvoli commented:
Pushed a new range controller test:

> **rac2: add dd test for msg app raft events**
> 
> Add a new `TestRangeController` variant `handle_raft_event_partial_msg_app`.
> 
> The test issues MsgApps in addition to the usual entries e.g.
> 
> ```
> raft_event
> range_id=1
>   entries
>     term=1 index=1 pri=NormalPri size=1MiB
>     term=1 index=2 pri=NormalPri size=1MiB
>     term=1 index=3 pri=NormalPri size=1MiB
>   sending
>     replica_id=1 [1,4)
>     replica_id=2 [1,3)
>     replica_id=3 [1,2)
> ```
> Part of: cockroachdb#130432
> Release note: None

kvoli commented:
The unit test failure is unrelated. I'm going to rebase the PR on master to hopefully resolve it.

sumeerbhola reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 1 of 1 files at r12, 1 of 1 files at r13, 2 of 2 files at r14.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/130908#-:-O7a2n8n84bBprTZOsPS:bg7igkv)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @kvoli and @pav-kv)

___
*[`pkg/kv/kvserver/kvflowcontrol/rac2/range_controller.go` line 507 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/130908#-O7a-rsUE5lDHvFIHupX:-O7a-rsUE5lDHvFIHupY:b3hlaip) ([raw file](https://github.com/cockroachdb/cockroach/blob/08bac17798002e5bc8ad408191c7f190a62d87eb/pkg/kv/kvserver/kvflowcontrol/rac2/range_controller.go#L507)):*
> ```Go
> 	if n := len(raftEventAppendState.newEntries); n > 0 {
> 		firstNewEntryIndex = raftEventAppendState.newEntries[0].index
> 		lastNewEntryIndex = raftEventAppendState.newEntries[n-1].index + 1
> ```

related to the comment in the test, we can add an assertion here that `latestFollowerStateInfo.Next <= lastNewEntryIndex`.

And for the else case, i.e., `n == 0` assert that `latestFollowerStateInfo.Next <= raftEventAppendState.rewoundNextRaftIndex`.
___
*[`pkg/kv/kvserver/kvflowcontrol/rac2/testdata/range_controller/handle_raft_event_partial_msg_app` line 58 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/130908#-O7a1OVi1b-jdPkrknq2:-O7a1OVi1b-jdPkrknq3:bl8hxay) ([raw file](https://github.com/cockroachdb/cockroach/blob/08bac17798002e5bc8ad408191c7f190a62d87eb/pkg/kv/kvserver/kvflowcontrol/rac2/testdata/range_controller/handle_raft_event_partial_msg_app#L58)):*
> ```
>   sending
>     replica_id=1 [4,6)
>     replica_id=2 [2,5)
> ```

are we recreating replica 2 and 3 here because of the regression? If yes, can you add a comment explaining this case.
___
*[`pkg/kv/kvserver/kvflowcontrol/rac2/testdata/range_controller/handle_raft_event_partial_msg_app` line 108 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/130908#-O7a1tkbDNfVSksJ1lzf:-O7a1tkbDNfVSksJ1lzg:b-rrw1dj) ([raw file](https://github.com/cockroachdb/cockroach/blob/08bac17798002e5bc8ad408191c7f190a62d87eb/pkg/kv/kvserver/kvflowcontrol/rac2/testdata/range_controller/handle_raft_event_partial_msg_app#L108)):*
> ```
>   sending
>     replica_id=1 [6,7)
>     replica_id=2 [4,7)
> ```

is replica 2 being recreated?

If the stream_state output included [indexToSend, nextRaftIndex), i.e., the bounds of the send-queue, this would be easier to see.
___
*[`pkg/kv/kvserver/kvflowcontrol/rac2/range_controller_test.go` line 1572 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/130908#-O7_Z9Pl86k8Htm_4zKS:-O7_Z9Pl86k8Htm_4zKT:b-ua10cj) ([raw file](https://github.com/cockroachdb/cockroach/blob/50eeeaf9b5280fda923763322ffecf99ef88ca1f/pkg/kv/kvserver/kvflowcontrol/rac2/range_controller_test.go#L1572)):*
> ```Go
> 		// 			{index: 10, term: 1, usesFlowControl: true, tokens: 100, pri: raftpb.NormalPri},
> 		// 			{index: 11, term: 1, usesFlowControl: true, tokens: 200, pri: raftpb.NormalPri},
> 		// 			{index: 12, term: 1, usesFlowControl: true, tokens: 300, pri: raftpb.NormalPri},
> ```

was `newEntries` meant to contain index 12? The rest of the example, including the expected `newEntries` suggests only 10 and 11 were appended.
FollowerStateInfo.Next is 13 and is exclusive, so if we only have appended 10, 11, 12, this is ok. But if we have only appended 10, 11, it suggests something very wrong inside Raft.
___
*[`pkg/kv/kvserver/kvflowcontrol/rac2/range_controller_test.go` line 1627 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/130908#-O7_b77w2pSTH2VpMwD_:-O7_b77w2pSTH2VpMwDa:bwy1lxz) ([raw file](https://github.com/cockroachdb/cockroach/blob/50eeeaf9b5280fda923763322ffecf99ef88ca1f/pkg/kv/kvserver/kvflowcontrol/rac2/range_controller_test.go#L1627)):*
> ```Go
> 				State: tracker.StateReplicate,
> 				Match: 9,
> 				Next:  12,
> ```

This should never happen, in that rewoundNextRaftIndex plus newEntries should never lag Next.
We should add an assertion.


<!-- Sent from Reviewable.io -->

kvoli commented:
I'll come back to this PR tomorrow morning. Feel free to address any of the test review comments before, or I'll get to them then.

kvoli committed:
rac2: include [match,indexToSend) and [indexToSend,nextRaftIndex)

Include `[match,indexToSend)` and `[indexToSend,nextRaftIndex)` when
printing the `stream_state` in `TestRangeController`. These inform the
reader about the state of the send queue and inflight messages e.g.

```
(n1,s1):1: state=replicate closed=false inflight=[1,4) send_queue=[4,4)
NormalPri:
  term=1 index=1  tokens=1048576
  term=1 index=2  tokens=1048576
  term=1 index=3  tokens=1048576
```

Part of: #130432
Release note: None

kvoli reviewed (commented):
> Reviewed 3 of 4 files at r5, 1 of 4 files at r11.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/130908)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @pav-kv and @sumeerbhola)

___
*[`pkg/kv/kvserver/kvflowcontrol/rac2/range_controller.go` line 494 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/130908#-O714wXW5p1ZodJnlFAn:-O7edCmV1GRcuNsjjJEc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc3b0ea3fcab5349e7cf6f95003e6a24d8e3bd3a/pkg/kv/kvserver/kvflowcontrol/rac2/range_controller.go#L494)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

This function will need a dedicated test -- all the complexity in the PR is here.
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/kvflowcontrol/rac2/range_controller.go` line 524 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/130908#-O74ThD11i02DVSOocVo:-O75AULq7kGXHzw2Es4k:bb7ohn6) ([raw file](https://github.com/cockroachdb/cockroach/blob/e33f93f20859a5a7de8e3aec136a295830ac4d22/pkg/kv/kvserver/kvflowcontrol/rac2/range_controller.go#L524)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I'll leave this change to you when you take over this PR.
</blockquote></details>

sgtm
___
*[`pkg/kv/kvserver/kvflowcontrol/rac2/range_controller.go` line 507 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/130908#-O7a-rsUE5lDHvFIHupX:-O7e4Wqm44PqqRBenFvL:b-63yi5h) ([raw file](https://github.com/cockroachdb/cockroach/blob/08bac17798002e5bc8ad408191c7f190a62d87eb/pkg/kv/kvserver/kvflowcontrol/rac2/range_controller.go#L507)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

related to the comment in the test, we can add an assertion here that `latestFollowerStateInfo.Next <= lastNewEntryIndex`.

And for the else case, i.e., `n == 0` assert that `latestFollowerStateInfo.Next <= raftEventAppendState.rewoundNextRaftIndex`.
</blockquote></details>

Added these assertions.
___
*[`pkg/kv/kvserver/kvflowcontrol/rac2/testdata/range_controller/handle_raft_event_partial_msg_app` line 58 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/130908#-O7a1OVi1b-jdPkrknq2:-O7e8yMC6AuO3-XpEUC9:begh8gy) ([raw file](https://github.com/cockroachdb/cockroach/blob/08bac17798002e5bc8ad408191c7f190a62d87eb/pkg/kv/kvserver/kvflowcontrol/rac2/testdata/range_controller/handle_raft_event_partial_msg_app#L58)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

are we recreating replica 2 and 3 here because of the regression? If yes, can you add a comment explaining this case.
</blockquote></details>

Yes. Added.
___
*[`pkg/kv/kvserver/kvflowcontrol/rac2/testdata/range_controller/handle_raft_event_partial_msg_app` line 108 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/130908#-O7a1tkbDNfVSksJ1lzf:-O7eRQG36btE5VwE5um6:b-jypsd8) ([raw file](https://github.com/cockroachdb/cockroach/blob/08bac17798002e5bc8ad408191c7f190a62d87eb/pkg/kv/kvserver/kvflowcontrol/rac2/testdata/range_controller/handle_raft_event_partial_msg_app#L108)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

is replica 2 being recreated?

If the stream_state output included [indexToSend, nextRaftIndex), i.e., the bounds of the send-queue, this would be easier to see.
</blockquote></details>

It is, updated to include another commit w/ the bounds.
___
*[`pkg/kv/kvserver/kvflowcontrol/rac2/range_controller_test.go` line 1572 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/130908#-O7_Z9Pl86k8Htm_4zKS:-O7e2hlx-Ynul9uleJq0:b-l3kjlt) ([raw file](https://github.com/cockroachdb/cockroach/blob/50eeeaf9b5280fda923763322ffecf99ef88ca1f/pkg/kv/kvserver/kvflowcontrol/rac2/range_controller_test.go#L1572)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

was `newEntries` meant to contain index 12? The rest of the example, including the expected `newEntries` suggests only 10 and 11 were appended.
FollowerStateInfo.Next is 13 and is exclusive, so if we only have appended 10, 11, 12, this is ok. But if we have only appended 10, 11, it suggests something very wrong inside Raft.
</blockquote></details>

It wasn't here, I should have removed -- I included index 12 in newEntries to see if that would pass.
___
*[`pkg/kv/kvserver/kvflowcontrol/rac2/range_controller_test.go` line 1627 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/130908#-O7_b77w2pSTH2VpMwD_:-O7e3p8_E-cM5RzbLdiU:btueguy) ([raw file](https://github.com/cockroachdb/cockroach/blob/50eeeaf9b5280fda923763322ffecf99ef88ca1f/pkg/kv/kvserver/kvflowcontrol/rac2/range_controller_test.go#L1627)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

This should never happen, in that rewoundNextRaftIndex plus newEntries should never lag Next.
We should add an assertion.
</blockquote></details>

Added an assertion.


<!-- Sent from Reviewable.io -->

kvoli commented:
Diff for `flow_control_integration_v2/v1_to_v2_transition`:

<details><summary>Details</summary>
<p>

```diff
diff --git a/pkg/kv/kvserver/testdata/flow_control_integration_v2/v1_to_v2_transition b/pkg/kv/kvserver/testdata/flow_control_integration_v2/v1_to_v2_transition
index 51a5444f78b..50309275737 100644
--- a/pkg/kv/kvserver/testdata/flow_control_integration_v2/v1_to_v2_transition
+++ b/pkg/kv/kvserver/testdata/flow_control_integration_v2/v1_to_v2_transition
@@ -220,7 +220,7 @@ SELECT range_id, store_id, crdb_internal.humanize_bytes(total_tracked_tokens::IN
 
   range_id | store_id | total_tracked_tokens  
 -----------+----------+-----------------------
-  70       | 1        | 2.0 MiB               
+  70       | 1        | 0 B                   
   70       | 2        | 2.0 MiB               
   70       | 3        | 2.0 MiB               
 
@@ -236,7 +236,7 @@ ORDER BY name ASC;
 
   kvflowcontrol.tokens.eval.elastic.available    | 24 MiB   
   kvflowcontrol.tokens.eval.elastic.deducted     | 6.0 MiB  
-  kvflowcontrol.tokens.eval.elastic.disconnected | 0 B      
+  kvflowcontrol.tokens.eval.elastic.disconnected | 2.0 MiB  
   kvflowcontrol.tokens.eval.elastic.returned     | 6.0 MiB  
   kvflowcontrol.tokens.eval.elastic.unaccounted  | 0 B      
   kvflowcontrol.tokens.eval.regular.available    | 48 MiB   
@@ -246,7 +246,7 @@ ORDER BY name ASC;
   kvflowcontrol.tokens.eval.regular.unaccounted  | 0 B      
   kvflowcontrol.tokens.send.elastic.available    | 24 MiB   
   kvflowcontrol.tokens.send.elastic.deducted     | 6.0 MiB  
-  kvflowcontrol.tokens.send.elastic.disconnected | 0 B      
+  kvflowcontrol.tokens.send.elastic.disconnected | 2.0 MiB  
   kvflowcontrol.tokens.send.elastic.returned     | 6.0 MiB  
   kvflowcontrol.tokens.send.elastic.unaccounted  | 0 B      
   kvflowcontrol.tokens.send.regular.available    | 48 MiB   
@@ -283,8 +283,8 @@ ORDER BY name ASC;
 
   kvflowcontrol.tokens.eval.elastic.available    | 21 MiB   
   kvflowcontrol.tokens.eval.elastic.deducted     | 9.0 MiB  
-  kvflowcontrol.tokens.eval.elastic.disconnected | 0 B      
-  kvflowcontrol.tokens.eval.elastic.returned     | 6.0 MiB  
+  kvflowcontrol.tokens.eval.elastic.disconnected | 3.0 MiB  
+  kvflowcontrol.tokens.eval.elastic.returned     | 7.0 MiB  
   kvflowcontrol.tokens.eval.elastic.unaccounted  | 0 B      
   kvflowcontrol.tokens.eval.regular.available    | 48 MiB   
   kvflowcontrol.tokens.eval.regular.deducted     | 0 B      
@@ -293,8 +293,8 @@ ORDER BY name ASC;
   kvflowcontrol.tokens.eval.regular.unaccounted  | 0 B      
   kvflowcontrol.tokens.send.elastic.available    | 21 MiB   
   kvflowcontrol.tokens.send.elastic.deducted     | 9.0 MiB  
-  kvflowcontrol.tokens.send.elastic.disconnected | 0 B      
-  kvflowcontrol.tokens.send.elastic.returned     | 6.0 MiB  
+  kvflowcontrol.tokens.send.elastic.disconnected | 3.0 MiB  
+  kvflowcontrol.tokens.send.elastic.returned     | 7.0 MiB  
   kvflowcontrol.tokens.send.elastic.unaccounted  | 0 B      
   kvflowcontrol.tokens.send.regular.available    | 48 MiB   
   kvflowcontrol.tokens.send.regular.deducted     | 0 B      
@@ -315,7 +315,7 @@ ORDER BY name ASC;
 
   kvflowcontrol.tokens.eval.elastic.available    | 24 MiB   
   kvflowcontrol.tokens.eval.elastic.deducted     | 9.0 MiB  
-  kvflowcontrol.tokens.eval.elastic.disconnected | 3.0 MiB  
+  kvflowcontrol.tokens.eval.elastic.disconnected | 5.0 MiB  
   kvflowcontrol.tokens.eval.elastic.returned     | 9.0 MiB  
   kvflowcontrol.tokens.eval.elastic.unaccounted  | 0 B      
   kvflowcontrol.tokens.eval.regular.available    | 48 MiB   
@@ -325,7 +325,7 @@ ORDER BY name ASC;
   kvflowcontrol.tokens.eval.regular.unaccounted  | 0 B      
   kvflowcontrol.tokens.send.elastic.available    | 24 MiB   
   kvflowcontrol.tokens.send.elastic.deducted     | 9.0 MiB  
-  kvflowcontrol.tokens.send.elastic.disconnected | 3.0 MiB  
+  kvflowcontrol.tokens.send.elastic.disconnected | 5.0 MiB  
   kvflowcontrol.tokens.send.elastic.returned     | 9.0 MiB  
   kvflowcontrol.tokens.send.elastic.unaccounted  | 0 B      
   kvflowcontrol.tokens.send.regular.available    | 48 MiB   
@@ -534,20 +534,20 @@ SELECT name, crdb_internal.humanize_bytes(value::INT8)
    WHERE name LIKE '%kvflowcontrol%tokens%'
 ORDER BY name ASC;
 
-  kvflowcontrol.tokens.eval.elastic.available    | 21 MiB   
+  kvflowcontrol.tokens.eval.elastic.available    | 22 MiB   
   kvflowcontrol.tokens.eval.elastic.deducted     | 3.0 MiB  
-  kvflowcontrol.tokens.eval.elastic.disconnected | 0 B      
-  kvflowcontrol.tokens.eval.elastic.returned     | 0 B      
+  kvflowcontrol.tokens.eval.elastic.disconnected | 1.0 MiB  
+  kvflowcontrol.tokens.eval.elastic.returned     | 1.0 MiB  
   kvflowcontrol.tokens.eval.elastic.unaccounted  | 0 B      
   kvflowcontrol.tokens.eval.regular.available    | 48 MiB   
   kvflowcontrol.tokens.eval.regular.deducted     | 0 B      
   kvflowcontrol.tokens.eval.regular.disconnected | 0 B      
   kvflowcontrol.tokens.eval.regular.returned     | 0 B      
   kvflowcontrol.tokens.eval.regular.unaccounted  | 0 B      
-  kvflowcontrol.tokens.send.elastic.available    | 21 MiB   
+  kvflowcontrol.tokens.send.elastic.available    | 22 MiB   
   kvflowcontrol.tokens.send.elastic.deducted     | 3.0 MiB  
-  kvflowcontrol.tokens.send.elastic.disconnected | 0 B      
-  kvflowcontrol.tokens.send.elastic.returned     | 0 B      
+  kvflowcontrol.tokens.send.elastic.disconnected | 1.0 MiB  
+  kvflowcontrol.tokens.send.elastic.returned     | 1.0 MiB  
   kvflowcontrol.tokens.send.elastic.unaccounted  | 0 B      
   kvflowcontrol.tokens.send.regular.available    | 48 MiB   
   kvflowcontrol.tokens.send.regular.deducted     | 0 B      
@@ -573,7 +573,7 @@ ORDER BY name ASC;
 
   kvflowcontrol.tokens.eval.elastic.available    | 24 MiB   
   kvflowcontrol.tokens.eval.elastic.deducted     | 3.0 MiB  
-  kvflowcontrol.tokens.eval.elastic.disconnected | 0 B      
+  kvflowcontrol.tokens.eval.elastic.disconnected | 1.0 MiB  
   kvflowcontrol.tokens.eval.elastic.returned     | 3.0 MiB  
   kvflowcontrol.tokens.eval.elastic.unaccounted  | 0 B      
   kvflowcontrol.tokens.eval.regular.available    | 48 MiB   
@@ -583,7 +583,7 @@ ORDER BY name ASC;
   kvflowcontrol.tokens.eval.regular.unaccounted  | 0 B      
   kvflowcontrol.tokens.send.elastic.available    | 24 MiB   
   kvflowcontrol.tokens.send.elastic.deducted     | 3.0 MiB  
-  kvflowcontrol.tokens.send.elastic.disconnected | 0 B      
+  kvflowcontrol.tokens.send.elastic.disconnected | 1.0 MiB  
   kvflowcontrol.tokens.send.elastic.returned     | 3.0 MiB  
   kvflowcontrol.tokens.send.elastic.unaccounted  | 0 B      
   kvflowcontrol.tokens.send.regular.available    | 48 MiB   
@@ -622,7 +622,7 @@ SELECT range_id, store_id, crdb_internal.humanize_bytes(total_tracked_tokens::IN
   range_id | store_id | total_tracked_tokens  
 -----------+----------+-----------------------
   70       | 1        | 1.0 MiB               
-  70       | 2        | 1.0 MiB               
+  70       | 2        | 0 B                   
   70       | 3        | 1.0 MiB               
 
 
@@ -648,7 +648,7 @@ ORDER BY name ASC;
 
   kvflowcontrol.tokens.eval.elastic.available    | 24 MiB   
   kvflowcontrol.tokens.eval.elastic.deducted     | 6.0 MiB  
-  kvflowcontrol.tokens.eval.elastic.disconnected | 0 B      
+  kvflowcontrol.tokens.eval.elastic.disconnected | 2.0 MiB  
   kvflowcontrol.tokens.eval.elastic.returned     | 6.0 MiB  
   kvflowcontrol.tokens.eval.elastic.unaccounted  | 0 B      
   kvflowcontrol.tokens.eval.regular.available    | 48 MiB   
@@ -658,7 +658,7 @@ ORDER BY name ASC;
   kvflowcontrol.tokens.eval.regular.unaccounted  | 0 B      
   kvflowcontrol.tokens.send.elastic.available    | 24 MiB   
   kvflowcontrol.tokens.send.elastic.deducted     | 6.0 MiB  
-  kvflowcontrol.tokens.send.elastic.disconnected | 0 B      
+  kvflowcontrol.tokens.send.elastic.disconnected | 2.0 MiB  
   kvflowcontrol.tokens.send.elastic.returned     | 6.0 MiB  
   kvflowcontrol.tokens.send.elastic.unaccounted  | 0 B      
   kvflowcontrol.tokens.send.regular.available    | 48 MiB   
@@ -683,8 +683,8 @@ ORDER BY name ASC;
 
   kvflowcontrol.tokens.eval.elastic.available    | 21 MiB   
   kvflowcontrol.tokens.eval.elastic.deducted     | 9.0 MiB  
-  kvflowcontrol.tokens.eval.elastic.disconnected | 0 B      
-  kvflowcontrol.tokens.eval.elastic.returned     | 6.0 MiB  
+  kvflowcontrol.tokens.eval.elastic.disconnected | 2.0 MiB  
+  kvflowcontrol.tokens.eval.elastic.returned     | 7.0 MiB  
   kvflowcontrol.tokens.eval.elastic.unaccounted  | 0 B      
   kvflowcontrol.tokens.eval.regular.available    | 48 MiB   
   kvflowcontrol.tokens.eval.regular.deducted     | 0 B      
@@ -693,8 +693,8 @@ ORDER BY name ASC;
   kvflowcontrol.tokens.eval.regular.unaccounted  | 0 B      
   kvflowcontrol.tokens.send.elastic.available    | 21 MiB   
   kvflowcontrol.tokens.send.elastic.deducted     | 9.0 MiB  
-  kvflowcontrol.tokens.send.elastic.disconnected | 0 B      
-  kvflowcontrol.tokens.send.elastic.returned     | 6.0 MiB  
+  kvflowcontrol.tokens.send.elastic.disconnected | 2.0 MiB  
+  kvflowcontrol.tokens.send.elastic.returned     | 7.0 MiB  
   kvflowcontrol.tokens.send.elastic.unaccounted  | 0 B      
   kvflowcontrol.tokens.send.regular.available    | 48 MiB   
   kvflowcontrol.tokens.send.regular.deducted     | 0 B      
@@ -723,7 +723,7 @@ ORDER BY name ASC;
 
   kvflowcontrol.tokens.eval.elastic.available    | 24 MiB   
   kvflowcontrol.tokens.eval.elastic.deducted     | 9.0 MiB  
-  kvflowcontrol.tokens.eval.elastic.disconnected | 3.0 MiB  
+  kvflowcontrol.tokens.eval.elastic.disconnected | 5.0 MiB  
   kvflowcontrol.tokens.eval.elastic.returned     | 9.0 MiB  
   kvflowcontrol.tokens.eval.elastic.unaccounted  | 0 B      
   kvflowcontrol.tokens.eval.regular.available    | 48 MiB   
@@ -733,7 +733,7 @@ ORDER BY name ASC;
   kvflowcontrol.tokens.eval.regular.unaccounted  | 0 B      
   kvflowcontrol.tokens.send.elastic.available    | 24 MiB   
   kvflowcontrol.tokens.send.elastic.deducted     | 9.0 MiB  
-  kvflowcontrol.tokens.send.elastic.disconnected | 3.0 MiB  
+  kvflowcontrol.tokens.send.elastic.disconnected | 5.0 MiB  
   kvflowcontrol.tokens.send.elastic.returned     | 9.0 MiB  
   kvflowcontrol.tokens.send.elastic.unaccounted  | 0 B      
   kvflowcontrol.tokens.send.regular.available    | 48 MiB   
@@ -781,22 +781,22 @@ ORDER BY name ASC;
 
   kvflowcontrol.tokens.eval.elastic.available    | 24 MiB   
   kvflowcontrol.tokens.eval.elastic.deducted     | 12 MiB   
-  kvflowcontrol.tokens.eval.elastic.disconnected | 3.0 MiB  
+  kvflowcontrol.tokens.eval.elastic.disconnected | 6.0 MiB  
   kvflowcontrol.tokens.eval.elastic.returned     | 12 MiB   
   kvflowcontrol.tokens.eval.elastic.unaccounted  | 0 B      
   kvflowcontrol.tokens.eval.regular.available    | 48 MiB   
   kvflowcontrol.tokens.eval.regular.deducted     | 3.0 MiB  
-  kvflowcontrol.tokens.eval.regular.disconnected | 0 B      
+  kvflowcontrol.tokens.eval.regular.disconnected | 1.0 MiB  
   kvflowcontrol.tokens.eval.regular.returned     | 3.0 MiB  
   kvflowcontrol.tokens.eval.regular.unaccounted  | 0 B      
   kvflowcontrol.tokens.send.elastic.available    | 24 MiB   
   kvflowcontrol.tokens.send.elastic.deducted     | 12 MiB   
-  kvflowcontrol.tokens.send.elastic.disconnected | 3.0 MiB  
+  kvflowcontrol.tokens.send.elastic.disconnected | 6.0 MiB  
   kvflowcontrol.tokens.send.elastic.returned     | 12 MiB   
   kvflowcontrol.tokens.send.elastic.unaccounted  | 0 B      
   kvflowcontrol.tokens.send.regular.available    | 48 MiB   
   kvflowcontrol.tokens.send.regular.deducted     | 3.0 MiB  
-  kvflowcontrol.tokens.send.regular.disconnected | 0 B      
+  kvflowcontrol.tokens.send.regular.disconnected | 1.0 MiB  
   kvflowcontrol.tokens.send.regular.returned     | 3.0 MiB  
   kvflowcontrol.tokens.send.regular.unaccounted  | 0 B      
 
@@ -816,22 +816,22 @@ ORDER BY name ASC;
 
   kvflowcontrol.tokens.eval.elastic.available    | 21 MiB   
   kvflowcontrol.tokens.eval.elastic.deducted     | 15 MiB   
-  kvflowcontrol.tokens.eval.elastic.disconnected | 3.0 MiB  
+  kvflowcontrol.tokens.eval.elastic.disconnected | 6.0 MiB  
   kvflowcontrol.tokens.eval.elastic.returned     | 12 MiB   
   kvflowcontrol.tokens.eval.elastic.unaccounted  | 0 B      
   kvflowcontrol.tokens.eval.regular.available    | 45 MiB   
   kvflowcontrol.tokens.eval.regular.deducted     | 6.0 MiB  
-  kvflowcontrol.tokens.eval.regular.disconnected | 0 B      
+  kvflowcontrol.tokens.eval.regular.disconnected | 1.0 MiB  
   kvflowcontrol.tokens.eval.regular.returned     | 3.0 MiB  
   kvflowcontrol.tokens.eval.regular.unaccounted  | 0 B      
   kvflowcontrol.tokens.send.elastic.available    | 21 MiB   
   kvflowcontrol.tokens.send.elastic.deducted     | 15 MiB   
-  kvflowcontrol.tokens.send.elastic.disconnected | 3.0 MiB  
+  kvflowcontrol.tokens.send.elastic.disconnected | 6.0 MiB  
   kvflowcontrol.tokens.send.elastic.returned     | 12 MiB   
   kvflowcontrol.tokens.send.elastic.unaccounted  | 0 B      
   kvflowcontrol.tokens.send.regular.available    | 45 MiB   
   kvflowcontrol.tokens.send.regular.deducted     | 6.0 MiB  
-  kvflowcontrol.tokens.send.regular.disconnected | 0 B      
+  kvflowcontrol.tokens.send.regular.disconnected | 2.0 MiB  
   kvflowcontrol.tokens.send.regular.returned     | 3.0 MiB  
   kvflowcontrol.tokens.send.regular.unaccounted  | 0 B      
 
@@ -842,7 +842,7 @@ SELECT range_id, store_id, crdb_internal.humanize_bytes(total_tracked_tokens::IN
 
   range_id | store_id | total_tracked_tokens  
 -----------+----------+-----------------------
-  70       | 1        | 1.0 MiB               
+  70       | 1        | 0 B                   
   70       | 2        | 1.0 MiB               
   70       | 3        | 1.0 MiB               
 
@@ -876,22 +876,22 @@ ORDER BY name ASC;
 
   kvflowcontrol.tokens.eval.elastic.available    | 24 MiB   
   kvflowcontrol.tokens.eval.elastic.deducted     | 21 MiB   
-  kvflowcontrol.tokens.eval.elastic.disconnected | 3.0 MiB  
+  kvflowcontrol.tokens.eval.elastic.disconnected | 9.0 MiB  
   kvflowcontrol.tokens.eval.elastic.returned     | 21 MiB   
   kvflowcontrol.tokens.eval.elastic.unaccounted  | 0 B      
   kvflowcontrol.tokens.eval.regular.available    | 48 MiB   
   kvflowcontrol.tokens.eval.regular.deducted     | 12 MiB   
-  kvflowcontrol.tokens.eval.regular.disconnected | 0 B      
+  kvflowcontrol.tokens.eval.regular.disconnected | 4.0 MiB  
   kvflowcontrol.tokens.eval.regular.returned     | 12 MiB   
   kvflowcontrol.tokens.eval.regular.unaccounted  | 0 B      
   kvflowcontrol.tokens.send.elastic.available    | 24 MiB   
   kvflowcontrol.tokens.send.elastic.deducted     | 21 MiB   
-  kvflowcontrol.tokens.send.elastic.disconnected | 3.0 MiB  
+  kvflowcontrol.tokens.send.elastic.disconnected | 9.0 MiB  
   kvflowcontrol.tokens.send.elastic.returned     | 21 MiB   
   kvflowcontrol.tokens.send.elastic.unaccounted  | 0 B      
   kvflowcontrol.tokens.send.regular.available    | 48 MiB   
   kvflowcontrol.tokens.send.regular.deducted     | 12 MiB   
-  kvflowcontrol.tokens.send.regular.disconnected | 0 B      
+  kvflowcontrol.tokens.send.regular.disconnected | 4.0 MiB  
   kvflowcontrol.tokens.send.regular.returned     | 12 MiB   
   kvflowcontrol.tokens.send.regular.unaccounted  | 0 B      
 
@@ -925,22 +925,22 @@ ORDER BY name ASC;
 
   kvflowcontrol.tokens.eval.elastic.available    | 24 MiB   
   kvflowcontrol.tokens.eval.elastic.deducted     | 3.0 MiB  
-  kvflowcontrol.tokens.eval.elastic.disconnected | 0 B      
+  kvflowcontrol.tokens.eval.elastic.disconnected | 1.0 MiB  
   kvflowcontrol.tokens.eval.elastic.returned     | 3.0 MiB  
   kvflowcontrol.tokens.eval.elastic.unaccounted  | 0 B      
   kvflowcontrol.tokens.eval.regular.available    | 48 MiB   
   kvflowcontrol.tokens.eval.regular.deducted     | 3.0 MiB  
-  kvflowcontrol.tokens.eval.regular.disconnected | 0 B      
+  kvflowcontrol.tokens.eval.regular.disconnected | 1.0 MiB  
   kvflowcontrol.tokens.eval.regular.returned     | 3.0 MiB  
   kvflowcontrol.tokens.eval.regular.unaccounted  | 0 B      
   kvflowcontrol.tokens.send.elastic.available    | 24 MiB   
   kvflowcontrol.tokens.send.elastic.deducted     | 3.0 MiB  
-  kvflowcontrol.tokens.send.elastic.disconnected | 0 B      
+  kvflowcontrol.tokens.send.elastic.disconnected | 1.0 MiB  
   kvflowcontrol.tokens.send.elastic.returned     | 3.0 MiB  
   kvflowcontrol.tokens.send.elastic.unaccounted  | 0 B      
   kvflowcontrol.tokens.send.regular.available    | 48 MiB   
   kvflowcontrol.tokens.send.regular.deducted     | 3.0 MiB  
-  kvflowcontrol.tokens.send.regular.disconnected | 0 B      
+  kvflowcontrol.tokens.send.regular.disconnected | 1.0 MiB  
   kvflowcontrol.tokens.send.regular.returned     | 3.0 MiB  
   kvflowcontrol.tokens.send.regular.unaccounted  | 0 B      
 
@@ -966,22 +966,22 @@ ORDER BY name ASC;
 
   kvflowcontrol.tokens.eval.elastic.available    | 24 MiB   
   kvflowcontrol.tokens.eval.elastic.deducted     | 21 MiB   
-  kvflowcontrol.tokens.eval.elastic.disconnected | 3.0 MiB  
+  kvflowcontrol.tokens.eval.elastic.disconnected | 9.0 MiB  
   kvflowcontrol.tokens.eval.elastic.returned     | 21 MiB   
   kvflowcontrol.tokens.eval.elastic.unaccounted  | 0 B      
   kvflowcontrol.tokens.eval.regular.available    | 48 MiB   
   kvflowcontrol.tokens.eval.regular.deducted     | 12 MiB   
-  kvflowcontrol.tokens.eval.regular.disconnected | 0 B      
+  kvflowcontrol.tokens.eval.regular.disconnected | 4.0 MiB  
   kvflowcontrol.tokens.eval.regular.returned     | 12 MiB   
   kvflowcontrol.tokens.eval.regular.unaccounted  | 0 B      
   kvflowcontrol.tokens.send.elastic.available    | 24 MiB   
   kvflowcontrol.tokens.send.elastic.deducted     | 21 MiB   
-  kvflowcontrol.tokens.send.elastic.disconnected | 3.0 MiB  
+  kvflowcontrol.tokens.send.elastic.disconnected | 9.0 MiB  
   kvflowcontrol.tokens.send.elastic.returned     | 21 MiB   
   kvflowcontrol.tokens.send.elastic.unaccounted  | 0 B      
   kvflowcontrol.tokens.send.regular.available    | 48 MiB   
   kvflowcontrol.tokens.send.regular.deducted     | 12 MiB   
-  kvflowcontrol.tokens.send.regular.disconnected | 0 B      
+  kvflowcontrol.tokens.send.regular.disconnected | 4.0 MiB  
   kvflowcontrol.tokens.send.regular.returned     | 12 MiB   
   kvflowcontrol.tokens.send.regular.unaccounted  | 0 B      
 
@@ -1007,7 +1007,7 @@ ORDER BY name ASC;
 
   kvflowcontrol.tokens.eval.elastic.available    | 24 MiB   
   kvflowcontrol.tokens.eval.elastic.deducted     | 9.0 MiB  
-  kvflowcontrol.tokens.eval.elastic.disconnected | 3.0 MiB  
+  kvflowcontrol.tokens.eval.elastic.disconnected | 5.0 MiB  
   kvflowcontrol.tokens.eval.elastic.returned     | 9.0 MiB  
   kvflowcontrol.tokens.eval.elastic.unaccounted  | 0 B      
   kvflowcontrol.tokens.eval.regular.available    | 48 MiB   
@@ -1017,7 +1017,7 @@ ORDER BY name ASC;
   kvflowcontrol.tokens.eval.regular.unaccounted  | 0 B      
   kvflowcontrol.tokens.send.elastic.available    | 24 MiB   
   kvflowcontrol.tokens.send.elastic.deducted     | 9.0 MiB  
-  kvflowcontrol.tokens.send.elastic.disconnected | 3.0 MiB  
+  kvflowcontrol.tokens.send.elastic.disconnected | 5.0 MiB  
   kvflowcontrol.tokens.send.elastic.returned     | 9.0 MiB  
   kvflowcontrol.tokens.send.elastic.unaccounted  | 0 B      
   kvflowcontrol.tokens.send.regular.available    | 48 MiB   
```

</p>
</details>

kvoli commented:
Pushed the fix we discussed @sumeerbhola.

sumeerbhola reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 4 of 8 files at r2, 2 of 4 files at r5, 4 of 7 files at r15, 1 of 1 files at r17, 6 of 7 files at r18, 2 of 2 files at r19, 1 of 1 files at r22, 6 of 7 files at r23, 2 of 2 files at r24, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/130908#-:-O7f6IYT-KyS7FMeaDKC:bg7igkv)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @pav-kv)



<!-- Sent from Reviewable.io -->

kvoli commented:
bors r+
